### PR TITLE
Remove thread local objects and use new visitors

### PR DESCRIPTION
**Pull request by jun-he**
Follow up the discussion in #1139, this change remove thread local objects and use new visitors to keep the code straightforward.

### CODE REVIEW

File: api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java [hunk 1]
```diff
@@ -102,7 +102,15 @@ private boolean eval(ContentFile<?> file) {
       this.lowerBounds = file.lowerBounds();
       this.upperBounds = file.upperBounds();
 
-      return ExpressionVisitors.visitEvaluator(expr, this);
+      try {
+        return ExpressionVisitors.visitEvaluator(expr, this);
+      } finally {
+        // allow temporary state to be collected because this is in a thread-local
+        this.valueCounts = null;
```
Comment (rdsr) [inline api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java:109]: I think this is subtle and hard to understand on why this is needed. I prefer the the new visitor approach as you guys discussed in #1139.

Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java:109]: That sounds good to me.

Comment (jun-he) [reply api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java]: Thanks @rdsr and @rdblue for the comments. Updated the PR accordingly to remove the thread local objects and use new visitors to keep the code straightforward.

**Comment by rdblue** [comment]
Looks good to me. What do you think, @rdsr?

**Comment by rdsr** [comment]
+1. Looks great now!

**Comment by jun-he** [comment]
Thanks for the review!

**Pull request merged**